### PR TITLE
Allow general filter use instead of just median filters

### PR DESCRIPTION
**Pull request by keflavich**
Is there a nicer way to inherit the docstrings?

### CODE REVIEW

**Comment by keflavich** [comment]
Ready for review

This PR basically just allows other filters to be used.  I needed this for some reason I can't remember recently.... I think I wanted a percentile filter or something?

File: spectral_cube/dask_spectral_cube.py [hunk 1]
```diff
@@ -901,14 +901,24 @@ def spectral_smooth(array):
...
         """
-        Smooth the cube along the spectral dimension
+        Smooth the cube along the spectral dimension using a filter
 
         Parameters
         ----------
         ksize : int
             Size of the median filter (scipy.ndimage.filters.median_filter)
+        filter : function
+            A filter from scipy.ndimage.filters
```
Comment (e-koch) [inline spectral_cube/dask_spectral_cube.py:921]:
```suggestion
            A filter from `scipy.ndimage.filters <https://docs.scipy.org/doc/scipy/reference/ndimage.html#filters>`_.
```

File: spectral_cube/dask_spectral_cube.py [hunk 2]
```diff
@@ -901,14 +901,24 @@ def spectral_smooth(array):
...
+        return self.spectral_filter(ksize, filter=filter,
+                                    raise_error_jybm=raise_error_jybm,
+                                    **kwargs)
+
+    @add_save_to_tmp_dir_option
+    def spectral_filter(self, ksize, filter, raise_error_jybm=True,
+                        **kwargs):
         """
-        Smooth the cube along the spectral dimension
+        Smooth the cube along the spectral dimension using a filter
```
Comment (e-koch) [inline spectral_cube/dask_spectral_cube.py:914]:
```suggestion
        Smooth the cube along the spectral dimension using a scipy.ndimage filter.
```

File: spectral_cube/spectral_cube.py [hunk 3]
```diff
@@ -2701,14 +2701,42 @@ def reproject(self, header, order='bilinear', use_memmap=False,
...
         """
         Smooth the image in each spatial-spatial plane of the cube using a median filter.
 
         Parameters
         ----------
         ksize : int
-            Size of the median filter (scipy.ndimage.filters.median_filter)
+            Size of the median filter (scipy.ndimage.filters.*_filter)
+        filter : function
+            A filter from scipy.ndimage.filters
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py:2714]:
```suggestion
            A filter from scipy.ndimage.filters. The default is the median filter.
```

File: spectral_cube/spectral_cube.py [hunk 4]
```diff
@@ -2701,14 +2701,42 @@ def reproject(self, header, order='bilinear', use_memmap=False,
...
+    def spatial_filter(self, ksize, filter, update_function=None, raise_error_jybm=True, **kwargs):
+        """
+        Smooth the image in each spatial-spatial plane of the cube using a filter.
+
+        Parameters
+        ----------
+        ksize : int
+            Size of the median filter (scipy.ndimage.filters.*_filter)
+        filter : function
+            A filter from scipy.ndimage.filters
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py:2739]:
```suggestion
            A filter from `scipy.ndimage.filters <https://docs.scipy.org/doc/scipy/reference/ndimage.html#filters>`_.
```

File: spectral_cube/spectral_cube.py [hunk 5]
```diff
@@ -2701,14 +2701,42 @@ def reproject(self, header, order='bilinear', use_memmap=False,
...
+
+    @parallel_docstring
+    def spatial_filter(self, ksize, filter, update_function=None, raise_error_jybm=True, **kwargs):
+        """
+        Smooth the image in each spatial-spatial plane of the cube using a filter.
+
+        Parameters
+        ----------
+        ksize : int
+            Size of the median filter (scipy.ndimage.filters.*_filter)
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py:2737]:
```suggestion
            Size of the filter in pixels (scipy.ndimage.filters.*_filter).
```

File: spectral_cube/spectral_cube.py [hunk 6]
```diff
@@ -2701,14 +2701,42 @@ def reproject(self, header, order='bilinear', use_memmap=False,
...
+        """
+        return self.spatial_filter(ksize=ksize, filter=filter,
+                                   update_function=update_function,
+                                   raise_error_jybm=raise_error_jybm, **kwargs)
+
+
+    @parallel_docstring
+    def spatial_filter(self, ksize, filter, update_function=None, raise_error_jybm=True, **kwargs):
+        """
+        Smooth the image in each spatial-spatial plane of the cube using a filter.
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py:2732]:
```suggestion
        Smooth the image in each spatial-spatial plane of the cube using a scipy.ndimage filter.
```

File: spectral_cube/spectral_cube.py [hunk 7]
```diff
@@ -2767,11 +2795,38 @@ def _gsmooth_image(img, **kwargs):
 
         return newcube
 
+    @parallel_docstring
+    def spectral_filter(self, ksize, filter, use_memmap=True, verbose=0,
+            num_cores=None, **kwargs):
+        """
+        Smooth the cube along the spectral dimension using a filter
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py:2802]:
```suggestion
        Smooth the cube along the spectral dimension using a scipy.ndimage filter.
```

File: spectral_cube/spectral_cube.py [hunk 8]
```diff
@@ -2767,11 +2795,38 @@ def _gsmooth_image(img, **kwargs):
...
+
+        Parameters
+        ----------
+        ksize : int
+            Size of the median filter (scipy.ndimage.filters.median_filter)
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py:2807]:
```suggestion
            Size of the filter in spectral channels.
```

File: spectral_cube/spectral_cube.py [hunk 9]
```diff
@@ -2701,14 +2701,42 @@ def reproject(self, header, order='bilinear', use_memmap=False,
...
+    def spatial_smooth_median(self, ksize, update_function=None, raise_error_jybm=True,
+                              filter=ndimage.filters.median_filter, **kwargs):
         """
         Smooth the image in each spatial-spatial plane of the cube using a median filter.
 
         Parameters
         ----------
         ksize : int
-            Size of the median filter (scipy.ndimage.filters.median_filter)
+            Size of the median filter (scipy.ndimage.filters.*_filter)
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py:2712]:
```suggestion
            Size of the median filter in pixels (scipy.ndimage.filters.median_filter)
```

File: spectral_cube/dask_spectral_cube.py [hunk 10]
```diff
@@ -967,14 +977,16 @@ def convolve_wrapper(data, kernel=None, **kwargs):
...
     @add_save_to_tmp_dir_option
-    def spatial_smooth_median(self, ksize, raise_error_jybm=True, **kwargs):
+    def spatial_filter(self, ksize, filter, raise_error_jybm=True, **kwargs):
         """
         Smooth the image in each spatial-spatial plane of the cube using a median filter.
 
         Parameters
         ----------
         ksize : int
             Size of the median filter (scipy.ndimage.filters.median_filter)
```
Comment (e-koch) [inline spectral_cube/dask_spectral_cube.py:987]:
```suggestion
            Size of the filter in pixels.
```

File: spectral_cube/dask_spectral_cube.py [hunk 11]
```diff
@@ -967,14 +977,16 @@ def convolve_wrapper(data, kernel=None, **kwargs):
...
+        filter : function
+            A filter from scipy.ndimage.filters
```
Comment (e-koch) [inline spectral_cube/dask_spectral_cube.py:989]:
```suggestion
            A filter from `scipy.ndimage.filters <https://docs.scipy.org/doc/scipy/reference/ndimage.html#filters>`_.
```

File: spectral_cube/dask_spectral_cube.py [hunk 12]
```diff
@@ -901,14 +901,24 @@ def spectral_smooth(array):
...
+    def spectral_filter(self, ksize, filter, raise_error_jybm=True,
+                        **kwargs):
         """
-        Smooth the cube along the spectral dimension
+        Smooth the cube along the spectral dimension using a filter
 
         Parameters
         ----------
         ksize : int
             Size of the median filter (scipy.ndimage.filters.median_filter)
```
Comment (e-koch) [inline spectral_cube/dask_spectral_cube.py:919]:
```suggestion
            Size of the median filter in spectral channels (scipy.ndimage.filters.median_filter).
```

File: spectral_cube/tests/test_spectral_cube.py [hunk 13]
```diff
@@ -2467,6 +2467,23 @@ def test_spatial_smooth_median(data_adv, use_dask):
     np.testing.assert_almost_equal(cube_median[2].value, result2)
 
 
+@pytest.mark.parametrize('num_cores', (None, 1))
+def test_spectral_smooth_maxfilter(num_cores, data_adv, use_dask):
```
Comment (e-koch) [inline spectral_cube/tests/test_spectral_cube.py:2471]: We should have one more test like this for spatial smoothing.

**Review by e-koch — CHANGES_REQUESTED** [review]
Added some clarifying suggestions for the input parameters, and I think we should add a matching max filter test for spatial smoothing.

I'm fairly sure we don't need to loop over the whole range of ndimage filters; I think they should all return a ND array.

**Review by e-koch — APPROVED** [review]

**Pull request merged**